examples/mnist.jl

```diff
@@ -0,0 +1,128 @@
+using MNIST, ReverseDiff
+
+const BATCH_SIZE = 100
+const IMAGE_SIZE = 784
+const CLASS_COUNT = 10
+
+##################
+# data wrangling #
+##################
+
+# loading MNIST data #
+#--------------------#
+
+function load_mnist(data)
+    images, label_indices = data
+    labels = zeros(CLASS_COUNT, length(label_indices))
+    for i in eachindex(label_indices)
+        labels[Int(label_indices[i]) + 1, i] = 1.0
+    end
+    return images, labels
+end
+
+const TRAIN_IMAGES, TRAIN_LABELS = load_mnist(MNIST.traindata())
+const TEST_IMAGES, TEST_LABELS = load_mnist(MNIST.testdata())
+
+# loading batches #
+#-----------------#
+
+immutable Batch{W,B,P,L}
+    weights::W
+    bias::B
+    pixels::P
+    labels::L
+end
+
+function Batch(images, labels, i)
+    weights = zeros(CLASS_COUNT, IMAGE_SIZE)
+    bias = zeros(CLASS_COUNT)
+    range = i:(i + BATCH_SIZE - 1)
+    return Batch(weights, bias, images[:, range], labels[:, range])
+end
+
+function load_batch!(batch, images, labels, i)
+    offset = i - 1
+    for batch_col in 1:BATCH_SIZE
+        data_col = batch_col + offset
+        for k in 1:size(images, 1)
+            batch.pixels[k, batch_col] = images[k, data_col]
+        end
+        for k in 1:size(labels, 1)
+            batch.labels[k, batch_col] = labels[k, data_col]
+        end
+    end
+    return batch
+end
+
+####################
+# model definition #
+####################
+
+# Here, we make a model out of simple `softmax` and `cross_entropy` functions. This could be
+# improved by implementing something like Tensorflow's `softmax_cross_entropy_with_logits`,
+# but my main goal is to show off ReverseDiff rather than implement the best possible model.
+
+# objective definitions #
+#-----------------------#
+
+# Here we use `@forward` to tell ReverseDiff to differentiate this scalar function in
+# forward-mode. This allows us to call `minus_log.(y)` instead of `-(log.(y))`. By defining
+# our own "fused" `minus_log` kernel using `@forward`, the operation `minus_log.(y)` becomes
+# a single array instruction in the tape (instead of two separate ones,  as is the case with
+# `-(log.(y))`), buying us a slight performance gain. Note that we didn't *need* to do this;
+# it's simply a good place to show off the `@forward` feature.
+ReverseDiff.@forward minus_log(x::Real) = -log(x)
+
+cross_entropy(y′, y) = mean(sum(y′ .* (minus_log.(y)), 1))
+
+softmax(x) = (exp_x = exp.(x); exp_x ./ sum(exp_x))
```

Comment:
If the second dimension represents the `BATCH_SIZE` then it would be `sum(exp_x, 1)`. I'm just used to the first dimension being the `BATCH_SIZE` so I kind of just assumed that, should have taken a closer look. For each example/image we'd want to get a probability distribution of what class/label it thinks that image belongs to.